base/utferror.jl

```diff
@@ -0,0 +1,51 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+#=
+@doc """
+@brief      Error messages for Unicode / UTF support
+""" ->
+=#
+
+const UTF_ERR_SHORT = 1
+const UTF_ERR_CONT  = 2
+const UTF_ERR_LONG  = 3
+const UTF_ERR_NOT_LEAD = 4
+const UTF_ERR_NOT_TRAIL = 5
+const UTF_ERR_NOT_SURROGATE = 6
+const UTF_ERR_MISSING_SURROGATE = 7
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_16_TERMINATE = 10
+const UTF_ERR_NULL_32_TERMINATE = 11
+const UTF_ERR_MAX = 11
+
+const errMsgs = [
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>>) missing one or more continuation bytes)",
+    "invalid UTF-8 sequence starting at index <<1>> (0x<<2>> is not a continuation byte)",
+    "invalid UTF-8 sequence, overlong encoding starting at index <<1>> (0x<<2>>)",
+    "not a leading Unicode surrogate code unit at index <<1>> (0x<<2>>)",
+    "not a trailing Unicode surrogate code unit at index <<1>> (0x<<2>>)",
+    "not a valid Unicode surrogate code unit at index <<1>> (0x<<2>>",
+    "missing trailing Unicode surrogate code unit after index <<1>> (0x<<2>>)",
+    "invalid Unicode character starting at index <<1>> (0x<<2>> > 0x10ffff)",
+    "surrogate encoding not allowed in UTF-8 or UTF-32, at index <<1>> (0x<<2>>)",
+    "UTF16String data must be NULL-terminated",
+    "UTF32String data must be NULL-terminated"
+]
+#=
+@doc """
+@brief      Throws ArgumentError with information about the specific error, location, and character
+
+@param[in]  errcode Error code for Unicode error (one of UTF_ERR_*)
+@param[in]  charpos Index of invalid byte or character
+@param[in]  invchar Invalid byte or character
+
+@throws never returns, always throws ArgumentError
+""" ->
+=#
+function utf_errfunc(errcode::Integer, charpos, invchar)
```

Comment:
This current structure was motived by a number of things:
1) Long familiarity [25 years] with a very powerful system for having localizable messages,
    where the messages can handle differences in word order between languages, and can handle
    plain output messages, warning messages, and error messages, with any number of arguments
    included in the output.
2) Fairly severe performance problems with GC frames, when I first implemented this by adding
    a new UnicodeError.  If that were totally fixed (see @yuyichao's #11508), I would prefer to move
    back to using UnicodeError, but I really don't think that that should be a reason to hold this up.
3) Enum's not working in Base, at least not early enough to be useful for this.
    When I first started implementing this, I used Enums, which is what all of those const's really
    meant to be.
So, yes, these could be better, but not at the moment, because of issues out of my control.
I will definitely change them to be what I'd originally wanted (which I think you all as well would prefer), as soon as those issues are dealt with.